pkg/autodiscovery/providers/prometheus_common.go

```diff
@@ -0,0 +1,229 @@
+	Tags                          []string                    `mapstructure:"tags" json:"tags,omitempty"`
+	Service                       string                      `mapstructure:"service" json:"service,omitempty"`
+	MinCollectInterval            int                         `mapstructure:"min_collection_interval" json:"min_collection_interval,omitempty"`
+	EmptyDefaultHost              bool                        `mapstructure:"empty_default_hostname" json:"empty_default_hostname,omitempty"`
+}
+
+// LabelJoinsConfig contains the label join configuration fields
+type LabelJoinsConfig struct {
+	LabelsToMatch []string `mapstructure:"labels_to_match" json:"labels_to_match"`
+	LabelsToGet   []string `mapstructure:"labels_to_get" json:"labels_to_get"`
+}
+
+// ADConfig contains the autodiscovery configuration data for a PrometheusCheck
+type ADConfig struct {
+	ExcludeAutoconf    *bool     `mapstructure:"exclude_autoconfig_files"`
+	KubeAnnotations    *InclExcl `mapstructure:"kubernetes_annotations"`
+	KubeContainerNames []string  `mapstructure:"kubernetes_container_names"`
+	containersRe       *regexp.Regexp
+}
+
+// InclExcl contains the include/exclude data structure
+type InclExcl struct {
+	Incl map[string]string `mapstructure:"include"`
+	Excl map[string]string `mapstructure:"exclude"`
+}
+
+// defaultCheck has the default openmetrics check values
+// To be used when the checks configuration is empty
+var defaultCheck = &PrometheusCheck{
+	Instances: []*OpenmetricsInstance{
+		{
+			Metrics:   openmetricsDefaultMetrics,
+			Namespace: openmetricsDefaultNS,
+		},
+	},
+	AD: &ADConfig{
+		ExcludeAutoconf: boolPointer(true),
+		KubeAnnotations: &InclExcl{
+			Excl: map[string]string{prometheusScrapeAnnotation: "false"},
+			Incl: map[string]string{prometheusScrapeAnnotation: "true"},
+		},
+		KubeContainerNames: []string{},
+	},
+}
+
+// init prepares the PrometheusCheck structure and defaults its values
+func (pc *PrometheusCheck) init() {
+	pc.Do(func() {
+		pc.initInstances()
+		pc.initAD()
+	})
+}
+
+// initInstances defaults the Instances field in PrometheusCheck
+func (pc *PrometheusCheck) initInstances() {
+	if len(pc.Instances) == 0 {
+		// Put a default config
+		pc.Instances = append(pc.Instances, &OpenmetricsInstance{
+			Metrics:   openmetricsDefaultMetrics,
+			Namespace: openmetricsDefaultNS,
+		})
+		return
+	}
+
+	for _, instance := range pc.Instances {
+		// Default the required config values if not set
+		if len(instance.Metrics) == 0 {
+			instance.Metrics = openmetricsDefaultMetrics
+		}
+	}
+}
+
+// initAD defaults the AD field in PrometheusCheck
+// It also prepares the regex to match the containers by name
+func (pc *PrometheusCheck) initAD() {
+	if pc.AD == nil {
+		pc.AD = &ADConfig{}
+	}
+
+	pc.AD.defaultAD()
+	pc.AD.setContainersRegex()
+}
+
+// defaultAD defaults the values of the autodiscovery structure
+func (ad *ADConfig) defaultAD() {
+	if ad.ExcludeAutoconf == nil {
```

Comment:
What's this `ExcludeAutoconf` part, it does not seem to be used anywhere?

Reply:
this will be implemented in another PR just to limit the size if this PR (same for label selectors)